{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "codons = { 'UUU': 'Phe', 'UUC': 'Phe', 'uua': 'Leu', 'UUG': 'Leu',\n",
    "          'CUU': 'Leu', 'CUC': 'Leu', 'CUA': 'Leu', 'CUG': 'Leu',\n",
    "          'AUU': 'Ile', 'AUC': 'Ile', 'AUA': 'Ile', 'AUG': 'Met',\n",
    "          'GUU': 'Val', 'GUC': 'Val', 'GUA': 'Val', 'GUG': 'Val',\n",
    "          'UCU': 'Ser', 'UCC': 'Ser', 'UCA': 'Ser', 'UCG': 'Ser',\n",
    "          'CCU': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',\n",
    "          'ACU': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',\n",
    "          'GCU': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',\n",
    "          'UAU': 'Tyr', 'UAC': 'Tyr', 'UAA': 'STOP', 'UAG': 'STOP',\n",
    "          'CAU': 'His', 'CAC': 'His', 'CAA': 'Gln', 'CAG': 'Gln',\n",
    "          'AAU': 'Asn', 'AAC': 'Asn', 'AAA': 'Lys', 'AAG': 'Lys',\n",
    "          'GAU': 'Asp', 'GAC': 'Asp', 'GAA': 'Glu', 'GAG': 'Glu',\n",
    "          'UGU': 'Cys', 'UGC': 'Cys', 'UGA': 'STOP', 'UGG': 'Trp',\n",
    "          'CGU': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg',\n",
    "          'AGU': 'Ser', 'AGC': 'Ser', 'AGA': 'Arg', 'AGG': 'Arg',\n",
    "          'GGU': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly' }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqance = 'AUGUUUUUCCUAGCACGUGGCGGGAGGUCGUAA'\n",
    "DNA = 'ACGT'\n",
    "RNA = 'ACGU'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This Function is to validate if sequence is DNA or RNA or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DNAValidation(dnaseq, seqtype):\n",
    "    for n in dnaseq:\n",
    "        if n not in seqtype:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function is to count Neoclutides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def countingnuc(seq):\n",
    "    ns = {'A':0,\n",
    "          'T':0,\n",
    "          'C':0,\n",
    "          'G':0\n",
    "          }\n",
    "    for n in seq:\n",
    "        ns[n] = ns[n] +1\n",
    "    return ns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DNA to RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DNAtoRNA(seq):\n",
    "    return seq.replace('T', 'U')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Translating RNA to Protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#DNA--->RNA--->Amino Acid--->proteins\n",
    "def Translating(seq):\n",
    "    amino = ''\n",
    "    for i in range(0,len(seq),3):\n",
    "        if codons[seq[i:i+3]] != 'STOP':\n",
    "            amino = amino +\"-\"+codons[seq[i:i+3]]\n",
    "    return amino"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Point Mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def countMutation(s1,s2):\n",
    "    m = 0\n",
    "    for i in range(0,len(s2)):\n",
    "        if s1[i] == s2[i]:\n",
    "            continue\n",
    "        else:\n",
    "            m = m + 1\n",
    "    return m"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DNA complement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def comp(dnaseq):\n",
    "    c = ''\n",
    "    for i in dnaseq:\n",
    "        if i =='A':\n",
    "            c = c + 'T'\n",
    "        if i =='T':\n",
    "            c = c + 'A'\n",
    "        if i =='C':\n",
    "            c = c + 'G'\n",
    "        if i =='G':\n",
    "            c = c + 'C'\n",
    "    return c"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DNA Reverse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse(dnaseq):\n",
    "    c = ''\n",
    "    d = {\n",
    "        'A':'T',\n",
    "        'T':'A',\n",
    "        'C':'G',\n",
    "        'G':'C'\n",
    "    }\n",
    "    for n in dnaseq:\n",
    "        c = c + d[n]\n",
    "    \n",
    "    return c[::-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Inputs From User"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter The sequenece you need to validate :AUGUUUUUCCUAGCACGUGGCGGGAGGUCGUAA\n",
      "Enter The type sequenece you need to validate (DNA or RNA) :rna\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "Sequence_By_User = input(\"Enter The sequenece you need to validate :\")\n",
    "Type_of_validate = input(\"Enter The type sequenece you need to validate (DNA or RNA) :\").upper()\n",
    "if Type_of_validate=='DNA':\n",
    "    Type_of_validate=DNA\n",
    "else :\n",
    "    Type_of_validate=RNA\n",
    "print(DNAValidation(Sequence_By_User , Type_of_validate))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter The sequenece you need to count its nucleotides :ATGTTTTTCCT\n",
      "The number of nucleotides in this sequence = {'A': 1, 'T': 7, 'C': 2, 'G': 1}\n"
     ]
    }
   ],
   "source": [
    "countingnucleotides = input(\"Enter The DNA sequenece you need to count its nucleotides :\")\n",
    "print(f\"The number of nucleotides in this sequence = {countingnuc(countingnucleotides)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter The sequenece of DNA you want to convert it into RNA :ATGTTTTTCCT\n",
      "THe sequence after converted from DNA to RNA is AUGUUUUUCCU\n"
     ]
    }
   ],
   "source": [
    "DNATORNACONVERT = input(\"Enter The sequenece of DNA you want to convert it into RNA :\")\n",
    "print(f\"THe sequence after converted from DNA to RNA is {DNAtoRNA(DNATORNACONVERT)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter The sequenece of RNA you want to convert it into Protien :AUGUUUUUCCUAGCACGUGGCGGGAGGUCGUAA\n",
      "THe sequence after converted from RNA to Protien is -Met-Phe-Phe-Leu-Ala-Arg-Gly-Gly-Arg-Ser\n"
     ]
    }
   ],
   "source": [
    "RNATOPRotienCONVERT = input(\"Enter The sequenece of RNA you want to convert it into Protien :\")\n",
    "print(f\"THe sequence after converted from RNA to Protien is {Translating(RNATOPRotienCONVERT)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the The original Sequence before Mutation :AUGUUUUUCCUAGCACGUGGCGGGAGGUCGUAA\n",
      "Enter the The Mutated Sequence:AUCUUUUTCCUAGCACGUGGCGGGAGGUCGTAA\n",
      "The number of point mutation between the original sequence and mutated sequence = 3\n"
     ]
    }
   ],
   "source": [
    "Original_Sequence = input(\"Enter the The original Sequence before Mutation :\")\n",
    "Mutated_Sequence = input(\"Enter the The Mutated Sequence:\")\n",
    "print(f\"The number of point mutation between the original sequence and mutated sequence = {countMutation(Original_Sequence , Mutated_Sequence)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter The sequenece of The First Strand :GGGAAAATTTT\n",
      "The complementary strand is CCCTTTTAAAA\n"
     ]
    }
   ],
   "source": [
    "First_strand = input(\"Enter The sequenece of The First Strand :\")\n",
    "print(f\"The complementary strand is {comp(First_strand)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter The sequenece of The Strand you want to reverse it:GGGAAAATTTT\n",
      "The strand after reversed is AAAATTTTCCC\n"
     ]
    }
   ],
   "source": [
    "Reversed_Strand = input(\"Enter The sequenece of The Strand you want to reverse it:\")\n",
    "print(f\"The strand after reversed is {reverse(Reversed_Strand)}\")"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
